packages/xo-web/src/xo-app/new-vm/index.js

```diff
@@ -548,9 +548,18 @@ export default class NewVm extends BaseComponent {
     const defaultNetworkIds = this._getDefaultNetworkIds(template)
     forEach(template.VIFs, vifId => {
       const vif = getObject(storeState, vifId, resourceSet)
-      VIFs.push({
-        network: pool || isInResourceSet(vif.$network) ? vif.$network : defaultNetworkIds[0],
-      })
+      // insertion of the vif at the index that corresponds to the order chosen by the user when creating the VM
+      // this order is stored in the key named "device" of the VIF object.
+      // if the index is NaN, the insertion is done at the end of the array.
+      const vifIndex = Number(vif.device)
+      const networkObject = {
+        network: pool || isInResourceSet(vif.$network) ? vif.$network : defaultNetworkIds[0]
+      }
+      if (isNaN(vifIndex)) {
+        VIFs.push(networkObject)
+      } else {
+        VIFs.splice(vifIndex, 0, networkObject)
+      }
```

Comment:
I don't think it works as intended. If you add VIFs with devices that are bigger than the length of the array, it will push them at the end and you might end up with the wrong order:
```js
VIFs = []
VIFs.splice(3, 0, { device: 3 })
VIFs.splice(2, 0, { device: 2 })
VIFs.splice(1, 0, { device: 1 })
VIFs // [ { device: 3 }, { device: 1 }, { device: 2 } ]
```

Also, you should be able to use a `.sort()` to order them, right?

Reply:
I can do something like that:
```javascript
const VIFs = new Array(template.VIFs.length).fill(0)

VIFs.splice(2, 1, {device: 2})
VIFs.splice(1, 1, {device: 1})
VIFs.splice(0, 1, {device: 0})
VIFs // [ { device: 0 }, { device: 1 }, { device: 2 } ]
```
It should work fine, what do you think? 🤔

Reply:
It looks complicated, why not a simple sort with a custom comparator to put non-numeric values at the end?

Reply:
I can't think of an efficient way to use a ```.sort()``` that wouldn't require a second ```for``` loop 🤔

Reply:
I can't think of an efficient way to use a ```.sort()``` that whouldn't require a second for loop 🤔

Reply:
Why?

Reply:
I might be wrong but you need the entire array to sort it. Therefore we would have something like:
```javascript
let vifObjectArray = [];
forEach(template.VIFs, vifId => {
    const vif = getObject(storeState, vifId, resourceSet);
    vifObjectArray.push(vif);
})
vifObjectArray.sort((a, b) => Number(a.device) - Number(b.device));
for (vifObject of vifObjectArray) {
    VIFs.push(network: pool || isInResourceSet(vifObject.$network) ? vifObject.$network : defaultNetworkIds[0]);
)
```
Or am I missing something?